Utils/azure-explorer-common/src/com/microsoft/tooling/msservices/serviceexplorer/azure/webapp/deploymentslot/DeploymentSlotNode.java

```diff
@@ -32,24 +39,87 @@
     protected final String webAppId;
     protected final String slotName;
 
-
     public DeploymentSlotNode(final String slotId, final String webAppId, final DeploymentSlotModule parent,
-                              final String name, final String state, final String os,
-                              final String subscriptionId, final String hostName) {
-        super(slotId, name, LABEL, parent,subscriptionId, hostName, os, state);
+                              final String name, final String state, final String os, final String subscriptionId,
+                              final String hostName) {
+        super(slotId, name, LABEL, parent, subscriptionId, hostName, os, state);
         this.webAppId = webAppId;
         this.slotName = name;
         this.presenter = new DeploymentSlotNodePresenter();
         this.presenter.onAttachView(this);
     }
 
+    @Override
+    public List<NodeAction> getNodeActions() {
+        getNodeActionByName(ACTION_SWAP_WITH_PRODUCTION).setEnabled(this.state == WebAppBaseState.RUNNING);
+        return super.getNodeActions();
+    }
+
     @Override
     protected void loadActions() {
-        // todo
+        // todo: why only the stop action has icon?
+        addAction(ACTION_STOP, getIcon(this.os, this.label, WebAppBaseState.STOPPED),
+            createBackgroundActionListener("Stopping Deployment Slot", () -> stop()));
+        addAction(ACTION_START, createBackgroundActionListener("Starting Deployment Slot", () -> start()));
+        addAction(ACTION_RESTART,
+            createBackgroundActionListener("Restarting Deployment Slot", () -> restart()));
+        addAction(ACTION_SWAP_WITH_PRODUCTION,
+            createBackgroundActionListener("Swapping with Production", () -> swapWithProduction()));
+        addAction(ACTION_OPEN_IN_BROWSER, new NodeActionListener() {
+            @Override
+            protected void actionPerformed(NodeActionEvent e) {
+                DefaultLoader.getUIHelper().openInBrowser("http://" + hostName);
+            }
+        });
+
+        super.loadActions();
+    }
+
+    @Override
+    protected void onNodeClick(NodeActionEvent e) {
+        // RefreshableNode refresh itself whenever the node is clicked.
```

Comment:
I remember it only refreshes on the first time being clicked?

Reply:
Yes, verified it again. My fault. But isn't it a little weird for a single node refreshes its state on the first time being clicked? The state is supposed to be fetched just when the node is created. The behavior looks like this:

![comment](https://user-images.githubusercontent.com/5179063/46239213-f9b38480-c3c7-11e8-90dc-ee67cfeaea81.gif)